src/Language/PureScript/TypeChecker.hs

```diff
@@ -324,19 +324,31 @@ typeCheckAll moduleName _ = traverse go
       | otherwise = firstDuplicate xs
     firstDuplicate _ = Nothing
 
-  checkOrphanInstance :: Ident -> Qualified (ProperName 'ClassName) -> [Type] -> m ()
-  checkOrphanInstance dictName className@(Qualified (Just mn') _) tys'
-    | moduleName == mn' || any checkType tys' = return ()
+  checkOrphanInstance :: Ident -> Qualified (ProperName 'ClassName) -> TypeClassData -> [Type] -> m ()
+  checkOrphanInstance dictName className@(Qualified (Just mn') _) typeClass tys'
+    | moduleName == mn' || moduleName `S.member` nonOrphanModules = return ()
     | otherwise = throwError . errorMessage $ OrphanInstance dictName className tys'
     where
-    checkType :: Type -> Bool
-    checkType (TypeVar _) = False
-    checkType (TypeLevelString _) = False
-    checkType (TypeConstructor (Qualified (Just mn'') _)) = moduleName == mn''
-    checkType (TypeConstructor (Qualified Nothing _)) = internalError "Unqualified type name in checkOrphanInstance"
-    checkType (TypeApp t1 _) = checkType t1
-    checkType _ = internalError "Invalid type in instance in checkOrphanInstance"
-  checkOrphanInstance _ _ _ = internalError "Unqualified class name in checkOrphanInstance"
+    typeModule :: Type -> Maybe ModuleName
+    typeModule (TypeVar _) = Nothing
+    typeModule (TypeLevelString _) = Nothing
+    typeModule (TypeConstructor (Qualified (Just mn'') _)) = Just mn''
+    typeModule (TypeConstructor (Qualified Nothing _)) = internalError "Unqualified type name in checkOrphanInstance"
+    typeModule (TypeApp t1 _) = typeModule t1
+    typeModule _ = internalError "Invalid type in instance in checkOrphanInstance"
+
+    modulesByTypeIndex :: M.Map Int (Maybe ModuleName)
+    modulesByTypeIndex = M.fromList (zip [0 ..] (typeModule <$> tys'))
+
+    lookupModule :: Int -> S.Set ModuleName
+    lookupModule idx = case M.lookup idx modulesByTypeIndex of
+      Just ms -> S.fromList (toList ms)
+      Nothing -> internalError "Unknown type index in checkOrphanInstance"
+
+    nonOrphanModules :: S.Set ModuleName
```

Comment:
Could you please add a short comment explaining why we take the intersection here?